src/metabase/query_processor/middleware/update_used_cards.clj

```diff
@@ -12,14 +13,25 @@
 (set! *warn-on-reflection* true)
 
 (defn- update-used-cards!*
-  [used-card-ids]
-  (when (seq used-card-ids)
-    (qp.util/with-execute-async
-      (fn []
-        (try
-          (t2/update! :model/Card :id [:in used-card-ids] {:last_used_at :%now})
-          (catch Throwable e
-            (log/error e "Error updating used cards")))))))
+  [card-id-timestamps]
+  (let [card-id->timestamp (update-vals (group-by :id card-id-timestamps)
+                                        (fn [xs] (apply t/max (map :timestamp xs))))]
+    (log/debugf "Update last_used_at of %d cards" (count card-id->timestamp))
+    (try
+      (t2/update! :model/Card :id [:in (keys card-id->timestamp)]
+                  {:last_used_at (into [:case]
+                                       (apply concat (for [[id timestamp] card-id->timestamp]
+                                                       [[:= :id id] [:greatest [:coalesce :last_used_at (t/offset-date-time 0)] timestamp]])))})
```

Comment:
You do this `case` construction twice, and it's quite dense, I think it's worth breaking out a helper function. I would also prefer the slightly more efficient and less nested transducer version - `(into [:case] (mapcat ...) card-id->timestamp)`, but that doesn't matter so much.

Reply:
re: helper

wdyt is the best api for this helper? I gave this a shot with this

```clojure
(defn better-case
  [xs if-fn then-fn]
  (into [:case]
        (mapcat (fn [x]
                  [(if-fn x) (then-fn x)])
                xs)))

(better-case [[1 [1 2 3]] [2 [3 4 5]]]
             (fn [[_cnt ids]]
               [:in :id ids])
             (fn [[cnt _ids]]
               cnt))

```

but I don't think it has better readability.

Reply:
Oh I was thinking something much less general purpose:

```clojure(defn- set-by-case [val->ids]
  (into [:case]
        (mapcat (fn [[v ids]]
                  [[:in :id ids] v]))
        val->ids))

(set-by-case 
 {1 [1 2 3]
  2 [3 4 5]})
```

Reply:
And for this timestamp example, use a slightly different helper, and do some pre-processing:

```clojure
(defn set-by-case [id->v]
  (into [:case]
        (mapcat (fn [[id v]]
                  [[:= :id id] v]))
        id->v))

(set-by-case
 (update-vals card-id->timestamp
              (fn [t]
                [:greatest [:coalesce :last_used_at :start-of-time] t])))
```

Reply:
If you wanted a single helper you could swap the arg order in the first one, and check whether each is is a `seq`, but I think that's getting too magical.